### DISCUSSION TITLE
How to save trial responses to an array?

### DISCUSSION THREAD

**Discussion by chester108** [jspsych/jsPsych#1441] [Q&A] [answered]
Hello. I have the following trial:
```
var trial = {
  type: 'image-keyboard-response',
  stimulus: 'imgA.png',
  on_finish: function(data) {
    if(data.key_press == 85){
      data.correct = true;
    } else {
      data.correct = false;
    }
  }
};
```

I would like to save participants' responses to the on_finish function as an array (either true or false based on their answer). How should I do this? Any help is greatly appreciated!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @chester108, is that what you're looking for?
```javascript
var response_accuracy = [];
var trial = {
  type: 'image-keyboard-response',
  stimulus: 'imgA.png',
  on_finish: function(data) {
    var acc;
    if(data.key_press == 85){
      acc = true;
    } else {
      acc = false;
    }
    data.correct = acc;  // save response accuracy to the trial data
    response_accuracy.push(acc);  // add the response accuracy to the array
  }
};
```

**Reply by chester108** [reply]
@becky-gilbert thanks for replying! I will try this and confirm shortly. :)

**Reply by chester108** [reply]
@becky-gilbert this seems to be storing the responses, however, even when the response should be true, it is saying false. I'm unsure why.

**Reply by chester108** [reply]
this is my specific code:
```
var button_press = [];	
var trial_one = {
    type: 'image-button-response',
    stimulus: light_off,
    choices: ['PRESS'],
    trial_duration: 3000,
    on_finish: function(data){
    var press1;
        if(data.button_press== 0){
        press1= true;
    }
       else {
        press1=false;
    }
    data.correct=press1;
    button_press.push(press1);
}
}

var feedback_one = {
    type: "image-keyboard-response",
    trial_duration: 2000,
    stimulus:function(){
        var last_trial_correct=jsPsych.data.get().last(1).values()[0].correct;
        if (last_trial_correct){
            return jsPsych.randomization.sampleWithoutReplacement(button_pressed, 1); //Return 1 stimuli sampled from the randomized button pressed array without replacement
        } else {
            return jsPsych.randomization.sampleWithoutReplacement(button_notpressed, 1) //Return 1 stimuli sampled from the randomized button not pressed array without replacement
    }
    }
}
timeline.push(trial_one,feedback_one);
```

**Reply by becky-gilbert** [reply]
Hi @chester108, thanks for sharing your code. I think the main problem is that the image-button-response plugin stores the response as "button_pressed", not "button_press" (at least this is the case with the jsPsych version I just checked: v6.1.0). 

Another possible issue is that the button press value is stored as a string rather than a number, so you probably want to check the value against "0" instead of 0. The reason this might matter in your case is that, if a response is not made before the time limit, then the value of "button_pressed" will be `null`, and `null` is a [falsy value](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) that will match the number 0 when you use the `==` loose equality operator (but not when you use the [strict equality operator](https://medium.com/@tonero91/making-sense-of-javascript-loose-vs-strict-equality-7675b2b90b3c) `===`).

Anyway, try scoring the response like this:
```javascript
if(data.button_pressed==="0"){
  press1=true;
} else {
  press1=false;
}
```

**Reply by chester108** [reply]
oh wow @becky-gilbert you're amazing!! thank you for taking the time to do this. I really appreciate it.

**Reply by chester108** [reply]
@becky-gilbert this is an extension of the previous question. When I try to apply a similar code as you suggested yesterday to this trial:
```
var feedback=[];
var feedback_one = {
    type: "image-keyboard-response",
    trial_duration: 2000,
    stimulus:function(){
        var last_trial_correct=jsPsych.data.get().last(1).values()[0].correct;
        if (last_trial_correct){
            var feedback1=[];
            return feedback1=jsPsych.randomization.sampleWithReplacement(button_pressed, 1,[1,1]); //Return 1 stimuli sampled from the randomized button pressed array with replacement
        } else {
            return feedback1=jsPsych.randomization.sampleWithReplacement(button_notpressed, 1,[1,1])//Return 1 stimuli sampled from the randomized button not pressed array with replacement
    }
data.correct=feedback1;
feedback.push(feedback1);
```
I receive an error that the final lines are unreachable. When possible, can you please share how you'd save the randomly selected feedback response? Thank you :)

**Reply by becky-gilbert** [reply]
Please see [my responses](https://github.com/jspsych/jsPsych/discussions/1463#discussioncomment-336234) to your other post about this 😃
